From 51d99cd247ec46fd72f8d9e25a7cbe28366af6ca Mon Sep 17 00:00:00 2001
From: Jens Wiklander <jens.wiklander@linaro.org>
Date: Mon, 20 Apr 2026 15:59:04 +0200
Subject: [PATCH] core: fix cleanup_shm_refs() attribute masking

cleanup_shm_refs() doesn't use the OPTEE_MSG_ATTR_TYPE_MASK to correctly
mask out the OPTEE_MSG_ATTR_TYPE_* attribute types, leading to a
resource leak if, for instance, the OPTEE_MSG_ATTR_NONCONTIG bit is set
in the attribute. Fix this by adding the missing mask in the switch
statement.

Fixes: 85771856fe40 ("core: reference count registered shared memory")
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Reviewed-by: Jerome Forissier <jerome.forissier@arm.com>
---
 core/tee/entry_std.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/core/tee/entry_std.c b/core/tee/entry_std.c
index 1d786aaa..420aeba4 100644
--- a/core/tee/entry_std.c
+++ b/core/tee/entry_std.c
@@ -254,7 +254,7 @@ static void cleanup_shm_refs(const uint64_t *saved_attr,
 	size_t n;
 
 	for (n = 0; n < MIN((unsigned int)TEE_NUM_PARAMS, num_params); n++) {
-		switch (saved_attr[n]) {
+		switch (saved_attr[n] & OPTEE_MSG_ATTR_TYPE_MASK) {
 		case OPTEE_MSG_ATTR_TYPE_TMEM_INPUT:
 		case OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT:
 		case OPTEE_MSG_ATTR_TYPE_TMEM_INOUT:
-- 
2.43.0

